src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs

```diff
@@ -2087,6 +2095,67 @@ private BoundExpression MakeNullCheck(SyntaxNode syntax, BoundExpression rewritt
                     type);
         }
 
+        private BoundExpression RewriteBuiltInNativeShiftOperation(
+            BoundBinaryOperator? oldNode,
+            SyntaxNode syntax,
+            BinaryOperatorKind operatorKind,
+            BoundExpression loweredLeft,
+            BoundExpression loweredRight,
+            TypeSymbol type)
+        {
+            // For the predefined operators, the number of bits to shift in `x << count`, `x >> count` or `x >>> count` is computed as follows:
+            // [...]
+            // - When the type of `x` is `nint` or `nuint`, the shift count is given by
+            //   the low-order five bits of `count` on a 32 bit platform, or
+            //   the lower-order six bits of `count` on a 64 bit platform.
+            //   The shift count is computed as `count & (sizeof(nint) * 8 - 1)` or `count & (sizeof(nuint) * 8 - 1)`,
+            //   which is `count & 0x1F` on a 32 bits platform and `count & 0x3F` on a 64 bits platform.
+
+            Debug.Assert(loweredLeft.Type is { });
+            TypeSymbol leftType = loweredLeft.Type;
+            Debug.Assert(leftType.SpecialType is SpecialType.System_IntPtr or SpecialType.System_UIntPtr);
+
+            ConstantValue? rightConstantValue = loweredRight.ConstantValue;
+            Debug.Assert(loweredRight.Type is { });
+            TypeSymbol rightType = loweredRight.Type;
+            Debug.Assert(rightType.SpecialType == SpecialType.System_Int32);
+
+            var oldSyntax = _factory.Syntax;
+            _factory.Syntax = loweredRight.Syntax;
+
+            if (rightConstantValue != null
+                && rightConstantValue.Discriminator == ConstantValueTypeDiscriminator.Int32
+                && rightConstantValue.Int32Value is >= 0 and <= 0x1F)
```

Comment:
>rightConstantValue.Int32Value is >= 0 and <= 0x1F [](http://example.com/codeflow?start=19&length=49)

Perhaps we will be able to handle some negative numbers as well with the following logic?
```
int originalValue = rightConstantValue.Int32Value;
int maskedValue = originalValue & 0x1F;
bool useMaskedValue = (originalValue & ~((1 << 31) | maskedValue )) == 0;
```